### PR TITLE
Implement filtering users by Organization header

### PR DESCRIPTION
**Pull request by chrisvire**

### CODE REVIEW

File: source/OptimaJet.DWKit.StarterApplication/Controllers/UsersController.cs [hunk 1]
```diff
@@ -7,12 +7,14 @@
 using Microsoft.AspNetCore.Authentication.JwtBearer;
 using Microsoft.AspNetCore.Authorization;
 using Microsoft.AspNetCore.Mvc;
```
Comment (NullVoxPopuli) [inline source/OptimaJet.DWKit.StarterApplication/Controllers/UsersController.cs]: why are these here?

File: source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs [hunk 2]
```diff
@@ -6,17 +6,32 @@
...
             ) : base(loggerFactory, jsonApiContext, contextResolver)
         {
+            this.OrganizationContext = organizationContext;
+        }
+
+        public override IQueryable<User> Get()
+        {
+            if (OrganizationContext.InvalidOrganization) return base.Get().Take(0);
+            var users = base.Get().Include(u => u.OrganizationMemberships);
+            if (!OrganizationContext.HasOrganization) return users;
```
Comment (NullVoxPopuli) [inline source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs]: this'll return all users in the database, yeah? 
don't we want to scope it to users in all the organizations the current user is in?

File: source/OptimaJet.DWKit.StarterApplication/Services/HttpOrganizationContext.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+        public HttpOrganizationContext 
+        (
+            IHttpContextAccessor httpContextAccessor,
+            OrganizationService organizationService
+        )
+        {
+            this.HttpContext = httpContextAccessor.HttpContext;
+            this.OrganizationService = organizationService;
+            Update();
```
Comment (NullVoxPopuli) [inline source/OptimaJet.DWKit.StarterApplication/Services/HttpOrganizationContext.cs]: rather than calling update in the constructor, the methods / properties should be memoized.

(having side-effects in constructors makes testing more difficult)

File: source/OptimaJet.DWKit.StarterApplication/Controllers/UsersController.cs (hunk 1 above)
Comment (chrisvire) [reply source/OptimaJet.DWKit.StarterApplication/Controllers/UsersController.cs]: May have needed them at one time.  I hoped that Visual Studio would visually show unused namespaces.

File: source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs (hunk 2 above)
Comment (chrisvire) [reply source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs]: The current user can be in multiple organizations.  The organization switcher will select one organization.  Will it be able to select multiple organizations?

Where in the UI will be be viewing a list of users anyway?  Just curious.

File: source/OptimaJet.DWKit.StarterApplication/Services/HttpOrganizationContext.cs (hunk 3 above)
Comment (chrisvire) [reply source/OptimaJet.DWKit.StarterApplication/Services/HttpOrganizationContext.cs]: Explain how it should be different?

File: source/OptimaJet.DWKit.StarterApplication/Controllers/UsersController.cs (hunk 1 above)
Comment (chrisvire) [reply source/OptimaJet.DWKit.StarterApplication/Controllers/UsersController.cs]: All three of those namespaces are needed.

Comment (chrisvire) [reply source/OptimaJet.DWKit.StarterApplication/Controllers/UsersController.cs]: I submitted a commit that remove unnecessary using statements.

File: source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs (hunk 2 above)
Comment (NullVoxPopuli) [reply source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs]: the current requirements are 'one org' or 'all of the orgs a user belongs to'.

that'll be in @giancorzo 's PR: https://github.com/sillsdev/appbuilder-portal/pull/31

File: source/OptimaJet.DWKit.StarterApplication/Services/HttpOrganizationContext.cs (hunk 3 above)
Comment (NullVoxPopuli) [reply source/OptimaJet.DWKit.StarterApplication/Services/HttpOrganizationContext.cs]: for example, for the CurrentUser,
image

tihs defines a getter property that is memoized such that subsequent calls don't make db hits again.

the CurrentUser is stashed in the HttpContext, but could probably just be a private property on that class since scoped services last the lifetime of a request.

File: source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs [hunk 4]
```diff
@@ -6,17 +6,45 @@
...
             IDbContextResolver contextResolver
             ) : base(loggerFactory, jsonApiContext, contextResolver)
         {
+            this.OrganizationContext = organizationContext;
+            this.CurrentUserContext = currentUserContext;
+        }
+
+        public override IQueryable<User> Get()
+        {
+            if (OrganizationContext.InvalidOrganization) return base.Get().Take(0);
```
Comment (NullVoxPopuli) [inline source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs]: semantically, what is an InvalidOrganization? what does this property represent?

anywho, to return an empty collection, I think you can just do 
```cs
 Enumerable.Empty<type of p>().AsQueryable();
```

**Comment by chrisvire** [comment]
This is ready for review again.

File: source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs [hunk 5]
```diff
@@ -6,17 +6,45 @@
...
+            {
+                // No organization specified, so include all users in the all the organizations that the current user is a member
+                var currentUser = GetByAuth0Id(CurrentUserContext.Auth0Id).Result;
+                var currentUserOrgIds = currentUser.OrganizationMemberships.Select(o => o.OrganizationId);
+                return base.Get()
+                           .Include(u => u.OrganizationMemberships)
+                           .Where(u => u.OrganizationMemberships.Select(o => o.OrganizationId).Intersect(currentUserOrgIds).Any());
+            }
+            else
+            {
```
Comment (NullVoxPopuli) [inline source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs]: because we are returning in each block, we can omit the `else` part and reduce the number of indentations :)

File: source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs (hunk 4 above)
Comment (chrisvire) [reply source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs]: We tried that but got an exception (non async IQueryable?  Or something like that).  We put it back in and it worked.  Will change.

File: source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs (hunk 5 above)
Comment (chrisvire) [reply source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs]: OK

File: source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs (hunk 4 above)
Comment (chrisvire) [reply source/OptimaJet.DWKit.StarterApplication/Repositories/UserRepository.cs]: InvalidOrganization is that we parsed a OrgId (as an integer), but it wasn’t found.  So returning empty.  If you want this to be the same as having an empty value (returning users in the same orgs as current user), that would be fine with me.
